dask/dataframe/io/parquet/arrow.py

```diff
@@ -726,44 +515,73 @@ def read_partition(
             row_group = None
             partition_keys = None
         else:
             # `piece` contains (path, row_group, partition_keys)
-            (path, row_group, partition_keys) = piece
+            (path_or_frag, row_group, partition_keys) = piece
 
         if not isinstance(row_group, list):
             row_group = [row_group]
 
-        dfs = []
-        for rg in row_group:
-            piece = pq.ParquetDatasetPiece(
-                path,
-                row_group=rg,
-                partition_keys=partition_keys,
-                open_file_func=partial(fs.open, mode="rb"),
-            )
-            arrow_table = cls._parquet_piece_as_arrow(
-                piece, columns, partitions, **kwargs
-            )
-            df = cls._arrow_table_to_pandas(arrow_table, categories, **kwargs)
-
-            if len(row_group) > 1:
-                dfs.append(df)
-
-        if len(row_group) > 1:
-            df = pd.concat(dfs)
+        # Read in arrow table and convert to pandas
+        arrow_table = cls._read_table(
+            path_or_frag,
+            fs,
+            row_group,
+            columns,
+            schema,
+            filters,
+            partitions,
+            partition_keys,
+            **kwargs,
+        )
+        df = cls._arrow_table_to_pandas(arrow_table, categories, **kwargs)
+
+        # For pyarrow.dataset api, need to convert partition columns
+        # to categorigal manually for integer types.  Also, if we did
+        # not read directly from fragments, we will need to add the
+        # partitioned columns here.
+        if partitions and isinstance(partitions, list):
+            if not isinstance(path_or_frag, pa_ds.ParquetFileFragment):
+                keys_dict = {k: v for (k, v) in partition_keys}
+            else:
+                keys_dict = {}
+            for partition in partitions:
+                if partition.name in df.columns:
+                    # We read directly from fragments, so the partition
+                    # columns are already in our dataframe.  We just
+                    # need to convert non-categorical types.
+                    if df[partition.name].dtype != pd.Categorical:
+                        df[partition.name] = pd.Series(
+                            pd.Categorical(
+                                categories=partition.keys,
+                                values=df[partition.name].values,
+                            ),
+                            index=df.index,
+                        )
+                else:
+                    # We read from file paths, so the partition
+                    # columns are NOT in our dataframe yet.
+                    df[partition.name] = pd.Series(
+                        pd.Categorical(
+                            categories=partition.keys,
+                            values=[keys_dict.get(partition.name, None)] * len(df),
+                        ),
+                        index=df.index,
```

Comment:
Note, for my test, I commented this out this part, because this caused a big slowdown (not sure why I didn't see it before, and only now, though ..). 
This is a quite inefficient way to create a categorical column, taking up a lot of time. As comparison, the code in the python ParquetDataset that does the same is here: 

https://github.com/apache/arrow/blob/cf61fa1b8c1463f8f7c16e55c7275cac3111a0c2/python/pyarrow/parquet.py#L723-L748

So we might need to copy some logic from there (or do something similar using pandas instead of pyarrow. Also in pandas, if we know the index value, we can create a categorical more efficiently with `pd.Categorical.from_codes(np.full(len(table), index, dtype='i4'), categories=...)`

Reply:
Thanks for pointing this out :)  I have not been testing the "partitioned" case much on my end, so I completely forgot I was doing this.

Reply:
@rjzamora I think this comment should best be addressed before merging

Reply:
Whoops, sorry, I see now you already did this!

Reply:
Right - Sorry, I should have updated this thread.  The categorical column creation is now using pyarrow directly.